designs/2022-unused-disable-directive-flexible-config/README.md

```diff
@@ -0,0 +1,247 @@
+
+### New Design
+
+We propose a new "explicit severity" design where the user can set the severity level of unused disable directive reporting using either the CLI option or the config file option. This design has the following benefits:
+
+- The user should always be able to achieve the desired effect (`off` / `warn` / `error`), regardless of where they set the option from (CLI option, config file option, or even from a shareable config)
+- No confusing inconsistency where one of the options triggers errors but the other triggers warnings
+
+We will also change the default behavior to error on unused disable directives so that all users (unless they opt-out) will benefit from the detection (and automatic removal with `--fix`) of unused disable directives.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+How the new "explicit severity" design works:
+
+- The config and CLI options are both changed from booleans to instead accept standard severity level string values: `off`, `warn`, `error` (or a the corresponding number for each level)
+- If both options are present, the CLI option takes precedence over (overrides) the config option
+- If only one of the options is present, the value of the present option is used
+- If neither option is present, the default value (`error`) is used
+- Note: There's only one underlying setting, and the config and CLI options are just two different ways of controlling it
+
+To reduce complexity and avoid confusion, we will not allow boolean values to be provided anymore, except during a possible intermediary [phase](#phases).
+
+In addition to allowing an explicit severity level to be passed to the CLI option `--report-unused-disable-directives` (e.g. `--report-unused-disable-directives warn`), we will also allow the option to be provided without a value (e.g. `--report-unused-disable-directives`) (a common CLI shorthand) which will default to `error` severity, the same as today.
+
+### Phases
+
+The implementation of this RFC will likely involve two phases:
+
+1. Phase 1: A non-breaking change to add support for the new severity levels, in addition to the existing boolean values. This intermediary, dual-support phase will give shareable configs using `reportUnusedDisableDirectives: true` a chance to switch over to `reportUnusedDisableDirectives: "warn"` while supporting both the ESLint minor version (e.g. `v8.x`) in which phase 1 gets implemented as well as the ESLint major version in which phase 2 gets implemented (e.g. `v9.0`).
+2. Phase 2: A breaking change intended for a major release in which support for the boolean values is removed, and the default behavior is changed to `error`.
```

Comment:
Anything that doesn't cause a CI system to break is something that's not enforced at all in practice, so I'd expect that most users would rather have that outcome.

Reply:
> I'm not sure that everyone would want their CI system to break because of unused disable directives.

One of my goals with this RFC is to argue for increased strictness including that unused disable directives should trigger errors that break your lint build. So I'd be interested to gather a few more reactions to that still. But if there isn't enough support for that change (using `error`), then `warn` would be the next best thing, as long as we are okay with the downsides, including this [comment](https://github.com/eslint/eslint/discussions/16512#discussioncomment-4089769):

> Warnings were intended to be used temporarily. My belief is that using warnings long term results in people tuning them out as just console noise, so I don’t believe you are getting much benefit with long-term warnings.

---

As for this comment:

> I also think we should continue allowing `true` and `false` in config forever. There's no downside to leaving it, just as we have multiple options for `globals`.

I'm open to this, but I have some concerns:

1. `reportUnusedDisableDirectives: true` meaning `warn` is not intuitive in my opinion. People generally expect true/false to correspond to on/off, not some intermediary/halfway state.
2. `reportUnusedDisableDirectives: true` using `warn` is inconsistent with how `--report-unused-disable-directives` uses `error`. Addressing this inconsistency is one of this RFC's key goals.

So if we need to continue supporting booleans, then I would love to at least update `reportUnusedDisableDirectives: true` to mean `error`. Although, since this would of course still be a breaking change, it's worth asking again whether we might as well just remove the booleans altogether...

Reply:
I also think that `"warn"` should be default. That would still accomplish the goal of raising awareness of this feature, but without breaking our semver policies. For more strictness, user can set `"error"` in the config file. Setting `"error"` (or using a shareable config that sets `"error"`) would be considered an opt-in to the possibility that patch releases can break CI.

Reply:
> One of my goals with this RFC is to argue for increased strictness including that unused disable directives should trigger errors that break your lint build.

I understand your goal here, but I don't believe this is what most users would want or expect. How they use disable directives is a personal choice.

And yes, I know what I said about warnings. :) My thinking has evolved a bit while using some other tools, mostly rust-analyzer. While I did create warnings as a way to easily implement new rules without breaking CI builds, I've come to believe that there are other good use cases, and code problems that don't affect functionality seem like a good use of warnings to me.

> So if we need to continue supporting booleans, then I would love to at least update reportUnusedDisableDirectives: true to mean error. Although, since this would of course still be a breaking change, it's worth asking again whether we might as well just remove the booleans altogether...

Yeah, I think we want `true` to equal `"warn"` assuming that the new default is `"warn"`. We don't necessarily need to document using `true` and `false`. For example, we just have a paragraph at the end of the [globals configuration section](https://eslint.org/docs/latest/user-guide/configuring/language-options#specifying-globals) that says:

> For historical reasons, the boolean value false and the string value "readable" are equivalent to "readonly". Similarly, the boolean value true and the string value "writeable" are equivalent to "writable". However, the use of these older values is deprecated.

Reply:
Thanks for the feedback. I have updated the RFC to use `warn` as the default and to continue to allow the boolean values with no behavior change.